src/master/heartbeat_cluster.cpp

```diff
@@ -0,0 +1,647 @@
+    , ui_nodes (other.ui_nodes)
+    , ping_hosts (other.ping_hosts)
+  {
+    memcpy (&lock, &other.lock, sizeof (pthread_mutex_t));
+  }
+
+  cluster &
+  cluster::operator= (const cluster &other)
+  {
+    memcpy (&lock, &other.lock, sizeof (pthread_mutex_t));
+    sfd = other.sfd;
+    state = other.state;
+    group_id = other.group_id;
+    hostname = other.hostname;
+    nodes = other.nodes;
+    *myself = *other.myself;
+    *master = *other.master;
+    shutdown = other.shutdown;
+    hide_to_demote = other.hide_to_demote;
+    is_isolated = other.is_isolated;
+    is_ping_check_enabled = other.is_ping_check_enabled;
+    ui_nodes = other.ui_nodes;
+    ping_hosts = other.ping_hosts;
+
+    return *this;
+  }
+
+  cluster::~cluster ()
+  {
+    destroy ();
+  }
+
+  int
+  cluster::init ()
+  {
+    char hostname_cstr[MAXHOSTNAMELEN];
+    int error_code = GETHOSTNAME (hostname_cstr, MAXHOSTNAMELEN); // TODO CBRD-22864 replace with css_gethostname
+    if (error_code != NO_ERROR)
+      {
+	MASTER_ER_SET_WITH_OSERROR (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_BO_UNABLE_TO_FIND_HOSTNAME, 0);
+	return ER_BO_UNABLE_TO_FIND_HOSTNAME;
+      }
+
+    hostname.assign (hostname_cstr);
+    is_ping_check_enabled = true;
+
+    if (HA_GET_MODE () == HA_MODE_REPLICA)
+      {
+	state = node_entry::node_state::REPLICA;
+      }
+    else
+      {
+	state = node_entry::node_state::SLAVE;
+      }
+
+    error_code = init_nodes ();
+    if (error_code != NO_ERROR)
+      {
+	return error_code;
+      }
+
+    if (state == node_entry::node_state::REPLICA && myself != NULL)
+      {
+	MASTER_ER_LOG_DEBUG (ARG_FILE_LINE, "myself should be in the ha_replica_list\n");
+	return ER_FAILED;
+      }
+
+    error_code = init_replica_nodes ();
+    if (error_code != NO_ERROR)
+      {
+	return error_code;
+      }
+
+    if (myself == NULL)
+      {
+	MASTER_ER_LOG_DEBUG (ARG_FILE_LINE, "cannot find myself\n");
+	return ER_FAILED;
+      }
+    if (nodes.empty ())
+      {
+	MASTER_ER_LOG_DEBUG (ARG_FILE_LINE, "cluster cluster node list is empty\n");
+	MASTER_ER_SET (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_PRM_BAD_VALUE, 1, prm_get_name (PRM_ID_HA_NODE_LIST));
+	return ER_PRM_BAD_VALUE;
+      }
+
+    init_ping_hosts ();
+    bool valid_ping_host_exists = check_valid_ping_host ();
+    if (!valid_ping_host_exists)
+      {
+	return ER_FAILED;
+      }
+
+    return error_code;
+  }
+
+  void
+  cluster::destroy ()
+  {
+    for (const node_entry *node : nodes)
+      {
+	delete node;
+      }
+    nodes.clear ();
+
+    for (const ui_node *node : ui_nodes)
+      {
+	delete node;
+      }
+    ui_nodes.clear ();
+
+    for (const ping_host *host : ping_hosts)
+      {
+	delete host;
+      }
+    ping_hosts.clear ();
+  }
+
+  int
+  cluster::reload ()
+  {
+    int error_code = sysprm_reload_and_init (NULL, NULL);
+    if (error_code != NO_ERROR)
+      {
+	return error_code;
+      }
+
+    // save current state of the cluster
+    cluster copy = *this;
```

Comment:
Instead of making a copy and restoring copy in case of errors, maybe we can make a copy, work on copy and overwrite this only at the end when reload was successful.